File: src/rust/delta/src/inplace.rs
use std::cmp::Reverse;
use std::collections::BinaryHeap;

use crate::types::{Command, CyclePolicy, PlacedCommand};

// ── DFS color states for find_cycle_in_scc ──────────────────────────────────
const COLOR_UNVISITED: u8 = 0;
const COLOR_ON_PATH:   u8 = 1;
const COLOR_DONE:      u8 = 2;

/// Sentinel value meaning "vertex is in no non-trivial SCC".
const NO_SCC: usize = usize::MAX;

// ── Data structures ──────────────────────────────────────────────────────────

/// Source offset, destination offset, and length of one copy command.
#[derive(Clone, Copy)]
struct CopyInfo {
    src:    usize,
    dst:    usize,
    length: usize,
}

/// Non-trivial SCCs with per-SCC active counts and vertex-to-SCC mapping.
struct SccList {
    sccs:   Vec<Vec<usize>>,  // non-trivial SCCs only
    active: Vec<usize>,        // live member count per SCC
    id:     Vec<usize>,        // vertex → SCC index; NO_SCC = trivial
}

/// Statistics from in-place conversion.
#[derive(Debug, Default)]
pub struct InplaceStats {
    pub num_copies:       usize,
    pub num_adds:         usize,
    pub edges:            usize,
    pub cycles_broken:    usize,
    pub copies_converted: usize,
    pub bytes_converted:  usize,
}

// ── Tarjan SCC (unchanged) ───────────────────────────────────────────────────

/// Compute SCCs using iterative Tarjan's algorithm.
///
/// Returns SCCs in reverse topological order (sinks first); caller reverses
/// for source-first processing order.
///
/// R.E. Tarjan, "Depth-first search and linear graph algorithms,"
/// SIAM J. Comput., 1(2):146-160, June 1972.
fn tarjan_scc(adj: &[Vec<usize>], n: usize) -> Vec<Vec<usize>> {
    let mut index_counter = 0usize;
    let mut index = vec![usize::MAX; n]; // MAX = unvisited
    let mut lowlink = vec![0usize; n];
    let mut on_stack = vec![false; n];
    let mut tarjan_stack: Vec<usize> = Vec::new();
    let mut sccs: Vec<Vec<usize>> = Vec::new();
    // DFS call stack: (vertex, next_neighbor_index)
    let mut call_stack: Vec<(usize, usize)> = Vec::new();

    for start in 0..n {
        if index[start] != usize::MAX {
            continue;
        }

        index[start] = index_counter;
        lowlink[start] = index_counter;
        index_counter += 1;
        on_stack[start] = true;
        tarjan_stack.push(start);
        call_stack.push((start, 0));

        while let Some(&(v, ni)) = call_stack.last() {
            if ni < adj[v].len() {
                let w = adj[v][ni];
                call_stack.last_mut().unwrap().1 += 1;
                if index[w] == usize::MAX {
                    // Tree edge: descend into w
                    index[w] = index_counter;
                    lowlink[w] = index_counter;
                    index_counter += 1;
                    on_stack[w] = true;
                    tarjan_stack.push(w);
                    call_stack.push((w, 0));
                } else if on_stack[w] {
                    // Back-edge into current SCC
                    if index[w] < lowlink[v] {
                        lowlink[v] = index[w];
                    }
                }
            } else {
                call_stack.pop();
                if let Some(&(parent, _)) = call_stack.last() {
                    if lowlink[v] < lowlink[parent] {
                        lowlink[parent] = lowlink[v];
                    }
                }
                if lowlink[v] == index[v] {
                    let mut scc = Vec::new();
                    loop {
                        let w = tarjan_stack.pop().unwrap();
                        on_stack[w] = false;
                        scc.push(w);
                        if w == v {
                            break;
                        }
                    }
                    sccs.push(scc);
                }
            }
        }
    }

    sccs // sinks first; caller reverses for source-first order
}

// ── Cycle finder (unchanged interface, improved constants) ───────────────────

/// Find a cycle in the active subgraph of one SCC.
///
/// Designed for repeated calls within the same SCC between cycle breakings:
///
/// - `sid` / `scc_id`: replaces a scc_member[] bool array; O(1) per neighbor
///   check instead of O(|SCC|) set-then-clear per call.
/// - `color[]` is persistent across calls (COLOR_DONE entries from fully
///   explored vertices are not reset): total DFS work across all calls within
///   one SCC is O(|SCC| + E_SCC), not O(|SCC| × cycles_broken).
/// - `scan_start`: amortized outer-loop position; advances monotonically so
///   the total scan cost per SCC is O(|SCC|), not O(|SCC| × cycles_broken).
///
/// On cycle found: resets COLOR_ON_PATH path vertices to COLOR_UNVISITED so
/// they can be re-examined after victim removal; leaves COLOR_DONE intact.
/// On None: COLOR_DONE values persist; caller advances scc_ptr without cleanup.
fn find_cycle_in_scc(
    adj: &[Vec<usize>],
    scc: &[usize],
    sid: usize,
    scc_id: &[usize],
    removed: &[bool],
    color: &mut [u8],
    scan_start: &mut usize,
) -> Option<Vec<usize>> {
    let mut path: Vec<usize> = Vec::new();

    while *scan_start < scc.len() {
        let start = scc[*scan_start];
        if removed[start] || color[start] != COLOR_UNVISITED {
            *scan_start += 1;
            continue;
        }

        color[start] = COLOR_ON_PATH;
        path.push(start);
        let mut stack: Vec<(usize, usize)> = vec![(start, 0)];

        while !stack.is_empty() {
            let (v, ni) = *stack.last().unwrap();
            let mut next_ni = ni;
            let mut advanced = false;

            while next_ni < adj[v].len() {
                let w = adj[v][next_ni];
                next_ni += 1;
                if scc_id[w] != sid || removed[w] {
                    continue;
                }
                if color[w] == COLOR_ON_PATH {
                    // Back-edge: cycle found
                    let pos = path.iter().position(|&x| x == w).unwrap();
                    let cycle = path[pos..].to_vec();
                    for &u in &path {
                        color[u] = COLOR_UNVISITED;
                    }
                    return Some(cycle);
                }
                if color[w] == COLOR_UNVISITED {
                    stack.last_mut().unwrap().1 = next_ni;
                    color[w] = COLOR_ON_PATH;
                    path.push(w);
                    stack.push((w, 0));
                    advanced = true;
                    break;
                }
            }

            if !advanced {
                stack.pop();
                color[v] = COLOR_DONE; // Fully explored — persists across calls.
                path.pop();
            }
        }

        // start's entire reachable SCC-subgraph explored; no cycle.
        *scan_start += 1;
    }

    None
}

// ── Extracted helpers ────────────────────────────────────────────────────────

/// Build CRWI digraph on copy commands.
///
/// Edge i→j means copy i reads from a region that copy j will overwrite,
/// so i must execute before j.  O(n log n + E) sweep-line construction.
/// Returns (adj, in_deg).
fn build_crwi_digraph(
    copy_info: &[CopyInfo],
    n: usize,
    stats: &mut InplaceStats,
) -> (Vec<Vec<usize>>, Vec<usize>) {
    let mut adj:    Vec<Vec<usize>> = vec![Vec::new(); n];
    let mut in_deg: Vec<usize>      = vec![0; n];

    // Sort copy write-intervals by start; binary-search for each read interval.
    let mut write_sorted: Vec<usize> = (0..n).collect();
    write_sorted.sort_unstable_by_key(|&j| copy_info[j].dst);
    let write_starts: Vec<usize> = write_sorted.iter().map(|&j| copy_info[j].dst).collect();

    for i in 0..n {
        let src      = copy_info[i].src;
        let length   = copy_info[i].length;
        let read_end = src + length;
        // lo = first write with dst >= src; hi = first write with dst >= read_end.
        // Writes in [lo, hi) start inside [src, read_end) — they always overlap.
        // The write at lo-1 starts before src; overlaps iff its end exceeds src.
        let lo = write_starts.partition_point(|&ws| ws < src);
        let hi = write_starts.partition_point(|&ws| ws < read_end);
        if lo > 0 {
            let j = write_sorted[lo - 1];
            if j != i {
                let dj = copy_info[j].dst;
                let lj = copy_info[j].length;
                if dj + lj > src {
                    adj[i].push(j);
                    in_deg[j] += 1;
                    stats.edges += 1;
                }
            }
        }
        for k in lo..hi {
            let j = write_sorted[k];
            if j != i {
                adj[i].push(j);
                in_deg[j] += 1;
                stats.edges += 1;
            }
        }
    }
    (adj, in_deg)
}

/// Wrap tarjan_scc output into an SccList containing only non-trivial SCCs.
fn build_scc_list(adj: &[Vec<usize>], n: usize) -> SccList {
    let sccs_raw = tarjan_scc(adj, n);
    let mut id     = vec![NO_SCC; n];
    let mut sccs:   Vec<Vec<usize>> = Vec::new();
    let mut active: Vec<usize>      = Vec::new();

    for scc in &sccs_raw {
        if scc.len() > 1 {
            let sid = sccs.len();
            for &v in scc {
                id[v] = sid;
            }
            active.push(scc.len());
            sccs.push(scc.clone());
        }
    }
    SccList { sccs, active, id }
}

/// Select a victim copy to break a cycle when Kahn's algorithm stalls.
///
/// Constant: first remaining vertex.  Localmin: minimum-length copy in a cycle.
/// scc_ptr and scan_pos are advanced in place across repeated calls.
fn pick_victim(
    copy_info: &[CopyInfo],
    adj:       &[Vec<usize>],
    scc_list:  &mut SccList,
    removed:   &[bool],
    color:     &mut [u8],
    scc_ptr:   &mut usize,
    scan_pos:  &mut usize,
    policy:    CyclePolicy,
    n:         usize,
) -> usize {
    match policy {
        CyclePolicy::Constant => (0..n).find(|&i| !removed[i]).unwrap(),
        CyclePolicy::Localmin => loop {
            while *scc_ptr < scc_list.sccs.len() && scc_list.active[*scc_ptr] == 0 {
                *scc_ptr += 1;
                *scan_pos = 0;
            }
            if *scc_ptr >= scc_list.sccs.len() {
                // Safety fallback — should not happen with a correct graph.
                break (0..n).find(|&i| !removed[i]).unwrap();
            }
            let result = find_cycle_in_scc(
                adj,
                &scc_list.sccs[*scc_ptr],
                *scc_ptr,
                &scc_list.id,
                removed,
                color,
                scan_pos,
            );
            match result {
                Some(cycle) => {
                    break *cycle.iter().min_by_key(|&&i| (copy_info[i].length, i)).unwrap();
                }
                None => {
                    // This SCC's remaining subgraph is acyclic; advance.
                    *scc_ptr += 1;
                    *scan_pos = 0;
                }
            }
        },
    }
}

/// Run Kahn topological sort; when the heap stalls, call pick_victim to break
/// the cycle by materialising one copy as a literal add.
fn run_kahn(
    copy_info: &[CopyInfo],
    adj:       &[Vec<usize>],
    scc_list:  &mut SccList,
    in_deg:    &mut [usize],
    r:         &[u8],
    add_info:  &mut Vec<(usize, Vec<u8>)>,
    policy:    CyclePolicy,
    n:         usize,
    stats:     &mut InplaceStats,
) -> Vec<usize> {
    let mut removed    = vec![false; n];
    let mut topo_order = Vec::with_capacity(n);
    let mut color      = vec![COLOR_UNVISITED; n];
    let mut scc_ptr    = 0usize;
    let mut scan_pos   = 0usize;

    let mut heap: BinaryHeap<Reverse<(usize, usize)>> = BinaryHeap::new();
    for i in 0..n {
        if in_deg[i] == 0 {
            heap.push(Reverse((copy_info[i].length, i)));
        }
    }
    let mut processed = 0;

    while processed < n {
        // Drain all ready vertices.
        while let Some(Reverse((_, v))) = heap.pop() {
            if removed[v] {
                continue;
            }
            removed[v] = true;
            topo_order.push(v);
            processed += 1;
            if scc_list.id[v] != NO_SCC {
                scc_list.active[scc_list.id[v]] -= 1;
            }
            for &w in &adj[v] {
                if !removed[w] {
                    in_deg[w] -= 1;
                    if in_deg[w] == 0 {
                        heap.push(Reverse((copy_info[w].length, w)));
                    }
                }
            }
        }

        if processed >= n {
            break;
        }

        // Kahn stalled: all remaining vertices are in CRWI cycles.
        let victim = pick_victim(
            copy_info, adj, scc_list, &removed, &mut color,
            &mut scc_ptr, &mut scan_pos, policy, n,
        );
        let ci = copy_info[victim];
        add_info.push((ci.dst, r[ci.src..ci.src + ci.length].to_vec()));
        stats.cycles_broken    += 1;
        stats.copies_converted += 1;
        stats.bytes_converted  += ci.length;
        removed[victim] = true;
        processed += 1;
        if scc_list.id[victim] != NO_SCC {
            scc_list.active[scc_list.id[victim]] -= 1;
        }
        for &w in &adj[victim] {
            if !removed[w] {
                in_deg[w] -= 1;
                if in_deg[w] == 0 {
                    heap.push(Reverse((copy_info[w].length, w)));
                }
            }
        }
    }
    topo_order
}

// ── Public entry point ───────────────────────────────────────────────────────

/// Convert standard delta commands to in-place executable commands.
///
/// The returned commands can be applied to a buffer initialized with R
/// to reconstruct V in-place, without a separate output buffer.
///
/// Why overlaps don't always require add conversion:
/// When copy i reads from `[src_i, src_i+len_i)` and copy j writes to
/// `[dst_j, dst_j+len_j)`, and these intervals overlap, copy i MUST execute
/// before j overwrites its source data.  This ordering constraint is an edge
/// i→j in the CRWI (Copy-Read/Write-Intersection) digraph.  When the graph
/// is acyclic, a topological order gives a valid serial schedule — no
/// conversion needed.  A cycle i₁→i₂→…→iₖ→i₁ creates a circular dependency
/// with no valid schedule; breaking it materializes one copy as a literal add
/// (saving source bytes from R before the buffer is modified).
///
/// Algorithm (Burns, Long, Stockmeyer, IEEE TKDE 2003):
///   1. Annotate each command with its write offset in the output
///   2. Build CRWI digraph: edge i→j iff i's read interval intersects j's
///      write interval (Section 4.2)
///   3. Topological sort (Kahn); when the heap empties with nodes remaining,
///      a cycle exists — find it and convert the minimum-length copy to an add
///   4. Output: copies in topological order, then all adds
pub fn make_inplace(
    r: &[u8],
    commands: &[Command],
    policy: CyclePolicy,
) -> (Vec<PlacedCommand>, InplaceStats) {
    let mut stats = InplaceStats::default();

    if commands.is_empty() {
        return (Vec::new(), stats);
    }

    // Step 1: compute write offsets
    let mut copy_info: Vec<CopyInfo>       = Vec::new();
    let mut add_info:  Vec<(usize, Vec<u8>)> = Vec::new();
    let mut write_pos = 0usize;

    for cmd in commands {
        match cmd {
            Command::Copy { offset, length } => {
                copy_info.push(CopyInfo { src: *offset, dst: write_pos, length: *length });
                write_pos += length;
            }
            Command::Add { data } => {
                add_info.push((write_pos, data.clone()));
                write_pos += data.len();
            }
        }
    }

    let n = copy_info.len();
    if n == 0 {
        stats.num_adds = add_info.len();
        return (
            add_info
                .into_iter()
                .map(|(dst, data)| PlacedCommand::Add { dst, data })
                .collect(),
            stats,
        );
    }

    // Steps 2-3: build digraph, topological sort, break cycles
    let (adj, mut in_deg) = build_crwi_digraph(&copy_info, n, &mut stats);
    let mut scc_list      = build_scc_list(&adj, n);
    let topo_order        = run_kahn(
        &copy_info, &adj, &mut scc_list, &mut in_deg,
        r, &mut add_info, policy, n, &mut stats,
    );

    // Step 4: assemble result — copies in topo order, then all adds
    stats.num_copies = topo_order.len();
    let mut result: Vec<PlacedCommand> = Vec::new();
    for &i in &topo_order {
        let ci = copy_info[i];
        result.push(PlacedCommand::Copy { src: ci.src, dst: ci.dst, length: ci.length });
    }
    for (dst, data) in add_info {
        result.push(PlacedCommand::Add { dst, data });
    }
    stats.num_adds = result.len() - stats.num_copies;

    (result, stats)
}

File: src/haskell/Delta/Inplace.hs
{-# LANGUAGE BangPatterns #-}

module Delta.Inplace
  ( InplaceStats(..)
  , makeInplace
  ) where

import Data.Array (Array, (!), accumArray, bounds, listArray)
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import Data.Foldable (toList)
import Data.List (foldl', minimumBy, sortBy)
import Data.Ord (comparing)
import Data.Sequence (Seq, (|>))
import qualified Data.Sequence as Seq
import qualified Data.Graph as G
import qualified Data.IntMap.Strict as IM
import qualified Data.IntSet as IS
import qualified Data.Set as Set
import Delta.Types

data CopyInfo = CopyInfo
  { ciSrc :: !Int
  , ciDst :: !Int
  , ciLen :: !Int
  }
  deriving (Eq, Show)

data InplaceStats = InplaceStats
  { ipsNumCopies :: !Int
  , ipsNumAdds :: !Int
  , ipsEdges :: !Int
  , ipsCyclesBroken :: !Int
  , ipsCopiesConverted :: !Int
  , ipsBytesConverted :: !Int
  }
  deriving (Eq, Show)

emptyStats :: InplaceStats
emptyStats =
  InplaceStats
    { ipsNumCopies = 0
    , ipsNumAdds = 0
    , ipsEdges = 0
    , ipsCyclesBroken = 0
    , ipsCopiesConverted = 0
    , ipsBytesConverted = 0
    }

makeInplace :: ByteString -> [Command] -> CyclePolicy -> ([PlacedCommand], InplaceStats)
makeInplace ref commands policy
  | null commands = ([], emptyStats)
  | n == 0 =
      let adds = [PlacedAdd dst bytes | (dst, bytes) <- toList addInfo0]
          st = emptyStats {ipsNumAdds = length adds}
       in (adds, st)
  | otherwise =
      let (adj, indeg0, edgeCount) = buildCrwiDigraph copyArr n
          copyLens = IM.fromList [(i, ciLen (copyArr ! i)) | i <- [0 .. n - 1]]
          initialHeap =
            Set.fromList
              [ (copyLens IM.! i, i)
              | i <- [0 .. n - 1]
              , IM.findWithDefault 0 i indeg0 == 0
              ]
          (topoRev, addInfoFinal, stRun) =
            runKahn ref n adj indeg0 copyArr copyLens initialHeap policy addInfo0
          topo = reverse topoRev
          placedCopies =
            [ let c = copyArr ! i in PlacedCopy (ciSrc c) (ciDst c) (ciLen c)
            | i <- topo
            ]
          placedAdds = [PlacedAdd dst bytes | (dst, bytes) <- toList addInfoFinal]
          out = placedCopies ++ placedAdds
          st =
            stRun
              { ipsNumCopies = length placedCopies
              , ipsNumAdds = length placedAdds
              , ipsEdges = edgeCount
              }
       in (out, st)
  where
    (copyInfos, addInfo0) = collectInfo commands
    n = length copyInfos
    copyArr = listArray (0, n - 1) copyInfos

collectInfo :: [Command] -> ([CopyInfo], Seq (Int, ByteString))
collectInfo = go 0 [] Seq.empty
  where
    go !_ copyAcc addAcc [] = (reverse copyAcc, addAcc)
    go !dst copyAcc addAcc (cmd : rest) =
      case cmd of
        Copy src len ->
          go (dst + len) (CopyInfo src dst len : copyAcc) addAcc rest
        Add bytes ->
          let len = BS.length bytes
           in go (dst + len) copyAcc (addAcc |> (dst, bytes)) rest

-- O(n log n + E) CRWI build via interval sweep (matches Rust/C++ approach).
buildCrwiDigraph :: Array Int CopyInfo -> Int -> (IM.IntMap [Int], IM.IntMap Int, Int)
buildCrwiDigraph copyArr n =
  foldl' step (IM.empty, IM.fromList [(i, 0) | i <- idxs], 0) idxs
  where
    idxs = [0 .. n - 1]
    writeSortedList = sortBy (comparing (\j -> ciDst (copyArr ! j))) idxs
    writeSorted = listArray (0, n - 1) writeSortedList :: Array Int Int
    writeStarts = listArray (0, n - 1) [ciDst (copyArr ! (writeSorted ! k)) | k <- idxs] :: Array Int Int

    step (!adj, !indeg, !edges) i =
      let ci = copyArr ! i
          src = ciSrc ci
          readEnd = src + ciLen ci
          lo = lowerBound writeStarts n src
          hi = lowerBound writeStarts n readEnd
          back =
            if lo > 0
              then collectBack i src (lo - 1) []
              else []
          forward = [j | k <- [lo .. hi - 1], let j = writeSorted ! k, j /= i]
          outs = back ++ forward
          adj' =
            if null outs
              then adj
              else IM.insert i outs adj
          indeg' = foldl' (\d j -> IM.insertWith (+) j 1 d) indeg outs
          edges' = edges + length outs
       in (adj', indeg', edges')

    collectBack i src !k acc
      | k < 0 = reverse acc
      | otherwise =
          let j = writeSorted ! k
              cj = copyArr ! j
           in if not (overlapsLeft src cj)
                then reverse acc
                else
                  if j == i
                    then collectBack i src (k - 1) acc
                    else collectBack i src (k - 1) (j : acc)

    overlapsLeft src cj = ciDst cj + ciLen cj > src

lowerBound :: Array Int Int -> Int -> Int -> Int
lowerBound arr n target = go 0 n
  where
    go !lo !hi
      | lo >= hi = lo
      | otherwise =
          let mid = lo + (hi - lo) `div` 2
           in if arr ! mid < target
                then go (mid + 1) hi
                else go lo mid

runKahn
  :: ByteString
  -> Int
  -> IM.IntMap [Int]
  -> IM.IntMap Int
  -> Array Int CopyInfo
  -> IM.IntMap Int
  -> Set.Set (Int, Int)
  -> CyclePolicy
  -> Seq (Int, ByteString)
  -> ([Int], Seq (Int, ByteString), InplaceStats)
runKahn ref n adj indeg0 copyArr copyLens heap0 policy addInfo0 =
  go 0 IS.empty indeg0 heap0 [] addInfo0 emptyStats scc0
  where
    scc0 = buildSccState n adj

    go !processed !removed !indeg !heap !topoRev !addInfo !stats !scc
      | processed >= n = (topoRev, addInfo, stats)
      | not (Set.null heap) =
          let ((_, v), heap1) = Set.deleteFindMin heap
           in if IS.member v removed
                then go processed removed indeg heap1 topoRev addInfo stats scc
                else
                  let removed' = IS.insert v removed
                      scc' = markRemoved v scc
                      (indeg', heap2) = relaxOutEdges v removed' indeg heap1
                   in go (processed + 1) removed' indeg' heap2 (v : topoRev) addInfo stats scc'
      | otherwise =
          let (victim, scc1) = pickVictim policy n copyArr adj removed scc
              ci = copyArr ! victim
              literal = BS.take (ciLen ci) (BS.drop (ciSrc ci) ref)
              addInfo' = addInfo |> (ciDst ci, literal)
              removed' = IS.insert victim removed
              scc2 = markRemoved victim scc1
              (indeg', heap1) = relaxOutEdges victim removed' indeg heap
              stats' =
                stats
                  { ipsCyclesBroken = ipsCyclesBroken stats + 1
                  , ipsCopiesConverted = ipsCopiesConverted stats + 1
                  , ipsBytesConverted = ipsBytesConverted stats + ciLen ci
                  }
           in go (processed + 1) removed' indeg' heap1 topoRev addInfo' stats' scc2

    relaxOutEdges v removed indeg heap =
      foldl' step (indeg, heap) (IM.findWithDefault [] v adj)
      where
        step (!d, !h) w
          | IS.member w removed = (d, h)
          | otherwise =
              let old = IM.findWithDefault 0 w d
                  new = old - 1
                  d' = IM.insert w new d
                  h' = if new == 0 then Set.insert (copyLens IM.! w, w) h else h
               in (d', h')

data SccState = SccState
  { ssSccs :: !(Array Int (Array Int Int))
  , ssCount :: !Int
  , ssId :: !(Array Int Int)
  , ssActive :: !(IM.IntMap Int)
  , ssPtr :: !Int
  , ssScan :: !Int
  , ssColor :: !(IM.IntMap Int)
  }

buildSccState :: Int -> IM.IntMap [Int] -> SccState
buildSccState n adj =
  let sccsRaw = G.stronglyConnComp [(i, i, IM.findWithDefault [] i adj) | i <- [0 .. n - 1]]
      sccLists = [vs | G.CyclicSCC vs <- sccsRaw, length vs > 1]
      sccCount = length sccLists
      sccArr = listArray (0, sccCount - 1) (map toArray sccLists)
      sidArr =
        accumArray
          (\_ sid -> sid)
          (-1)
          (0, n - 1)
          [(v, sid) | (sid, vs) <- zip [0 ..] sccLists, v <- vs]
      active = IM.fromList [(sid, length vs) | (sid, vs) <- zip [0 ..] sccLists]
   in SccState
        { ssSccs = sccArr
        , ssCount = sccCount
        , ssId = sidArr
        , ssActive = active
        , ssPtr = 0
        , ssScan = 0
        , ssColor = IM.empty
        }
  where
    toArray vs = listArray (0, length vs - 1) vs

markRemoved :: Int -> SccState -> SccState
markRemoved v st =
  let sid = ssId st ! v
   in if sid < 0
        then st
        else st {ssActive = IM.adjust (\x -> x - 1) sid (ssActive st)}

pickVictim :: CyclePolicy -> Int -> Array Int CopyInfo -> IM.IntMap [Int] -> IS.IntSet -> SccState -> (Int, SccState)
pickVictim policy n copyArr adj removed st0 =
  case policy of
    Constant -> (firstRemaining, st0)
    Localmin -> choose st0
  where
    choose st
      | ssPtr st >= ssCount st = (firstRemaining, st)
      | IM.findWithDefault 0 (ssPtr st) (ssActive st) == 0 =
          choose st {ssPtr = ssPtr st + 1, ssScan = 0}
      | otherwise =
          let sid = ssPtr st
              sccVerts = ssSccs st ! sid
              (mCycle, color', scan') =
                findCycleInScc sid sccVerts (ssId st) adj removed (ssColor st) (ssScan st)
              st' = st {ssColor = color', ssScan = scan'}
           in case mCycle of
                Just cycleNodes -> (minimumBy (comparing key) cycleNodes, st')
                Nothing -> choose st' {ssPtr = sid + 1, ssScan = 0}

    firstRemaining =
      let go !i
            | i >= n = error "pickVictim: no remaining vertices"
            | IS.member i removed = go (i + 1)
            | otherwise = i
       in go 0

    key i = (ciLen (copyArr ! i), i)

findCycleInScc
  :: Int
  -> Array Int Int
  -> Array Int Int
  -> IM.IntMap [Int]
  -> IS.IntSet
  -> IM.IntMap Int
  -> Int
  -> (Maybe [Int], IM.IntMap Int, Int)
findCycleInScc sid sccVerts sccId adj removed color0 scan0 = goScan scan0 color0
  where
    !sccLen = let (_, hi) = bounds sccVerts in hi + 1

    goScan !scan color
      | scan >= sccLen = (Nothing, color, scan)
      | otherwise =
          let start = sccVerts ! scan
           in if IS.member start removed || colorAt start color /= 0
                then goScan (scan + 1) color
                else
                  case dfs start [start] color of
                    (Just cyc, color') -> (Just cyc, color', scan)
                    (Nothing, color') -> goScan (scan + 1) color'

    dfs v path color0 =
      let color1 = IM.insert v 1 color0
       in walk (IM.findWithDefault [] v adj) color1
      where
        walk [] color = (Nothing, IM.insert v 2 color)
        walk (w : ws) color
          | IS.member w removed || sccId ! w /= sid = walk ws color
          | otherwise =
              case colorAt w color of
                0 ->
                  let (res, color') = dfs w (w : path) color
                   in case res of
                        Just cyc -> (Just cyc, color')
                        Nothing -> walk ws color'
                1 ->
                  let seg = takeWhile (/= w) path
                      color' = clearPath path color
                   in (Just (reverse (w : seg)), color')
                _ -> walk ws color

    colorAt v color = IM.findWithDefault 0 v color
    clearPath path color = foldl' (flip IM.delete) color path
